platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,279 @@
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible for coordinating compaction tasks for a {@link MerkleDbDataSource}.
+ * It provides convenient API for starting compactions for each of the three storage types. Also, this class makes sure
+ * that there are no concurrent compactions for the same storage type. And finally it provides a way to stop all compactions
+ * and keep them disabled until they are explicitly enabled again.
+ * The compaction tasks are executed in a background thread pool.
+ * The number of threads in the pool is defined by {@link MerkleDbConfig#compactionThreads()} property.
+ *
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+class MerkleDbCompactionCoordinator {
+
+    /**
+     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
+     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
+     * holder will have been configured by the time this static initializer runs.
+     */
+    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+
+    /**
+     * A thread pool to run compaction tasks.
+     */
+    private static final AtomicReference<ExecutorService> compactionExecutorServiceRef = new AtomicReference<>();
+
+    private static final Logger logger = LogManager.getLogger(MerkleDbCompactionCoordinator.class);
+
+    public static final String HASH_STORE_DISK_SUFFIX = "HashStoreDisk";
+    public static final String OBJECT_KEY_TO_PATH_SUFFIX = "ObjectKeyToPath";
+    public static final String PATH_TO_KEY_VALUE_SUFFIX = "PathToKeyValue";
+    private final AtomicBoolean compactionEnabled = new AtomicBoolean();
+    // we need a map of exactly three elements, one per storage
+    final ConcurrentMap<String, InterruptibleCompletableFuture<?>> compactionFuturesByName = new ConcurrentHashMap<>(3);
+    private final MerkleDbStatisticsUpdater statisticsUpdater;
+    private final Optional<CompactionTask> objectKeyToPathTaskOpt;
+    private final Optional<CompactionTask> hashesStoreDiskTaskOpt;
+    private final CompactionTask pathToKeyValueTask;
+
+    /**
+     * Creates a new instance of {@link MerkleDbCompactionCoordinator}.
+     * @param tableName the name of the table
+     * @param statisticsUpdater a statistics updater
+     * @param objectKeyToPath an object key to path store
+     * @param hashesStoreDisk a hash store
+     * @param pathToKeyValue a path to key-value store
+     */
+    public MerkleDbCompactionCoordinator(
+            @NonNull String tableName,
+            @NonNull MerkleDbStatisticsUpdater statisticsUpdater,
+            @Nullable Compactible objectKeyToPath,
+            @Nullable Compactible hashesStoreDisk,
+            @NonNull Compactible pathToKeyValue) {
```

Comment:
I wonder if stat updates may be hidden inside compactibles. I mean, objectKeyToPath store knows it should update "leaf keys compaction time / saved space" stat, not "hashes compaction time / saved space". In this case, the coordinator will only have to deal with generic `Compactible` instances without any knowledge of what they actually compact. This constructor may accept an arbitrary list of compactibles, two, three, five, whatever number of stores we may have in the future

Reply:
Or it may be a map <String, Compactible>, task name to compactible object.

Reply:
Yes, it's a good idea to move the state it to `Compactible` implementations. I'll do it.